[Inductor/Triton] Customize triton codegen to optionally preserve input dtype on tl.load

arui-meta opened the pull request:
Differential Revision: D60536337


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

jfix71 reviewed (approved):
LGTM

blaine-rister commented on torch/_inductor/codegen/triton.py:
```diff
@@ -470,39 +474,39 @@ def _print_Abs(self, expr):
     def _print_OpaqueUnaryFn_cosh(self, expr):
         assert len(expr.args) == 1
-        return f"libdevice.cosh(({self._print(expr.args[0])}).to(tl.float32))"
+        return f"libdevice.cosh({self._print(expr.args[0])}{_maybe_upcast_to_fp32()})"
```
Do we know if all these `libdevice` calls support fp16? I recall @eellison  had some concerns about that.

eellison replied:
They dont. we're landing this to unblock mtia and doing more complete dtype analysis later

eellison replied:
although, cc @arui-meta, maybe we should we could just upcast input to fp32 and then downcast after and that would be sufficient ?

eellison replied:
As far as I can tell none of these libdevices support lower precision: https://github.com/triton-lang/triton/blob/14025786d108596cfd99700caa4f438938c2ceba/python/triton/language/extra/cuda/libdevice.py#L1192

blaine-rister replied:
I guess it makes sense to land this as an experimental feature to unblock MTIA. But I'm wondering if these casts for `libdevice` make sense on either arch. MTIA has its own `libdevice` which is different from GPU, but I don't know that it supports all these either.

arui-meta replied:
By default we still will upcast to FP32 for GPU, so GPU codegen behavior should stay the same. I'm afraid that we don't have any dtype information on the input args. We don't know what dtype we should downcast to.

eellison replied:
@arui-meta, you dont have to use compile time dtype, you can use runtime dtype `{expr.args[0]}.dtype`

arui-meta replied:
I have added upcast input to fp32 and then downcast to inputs dtype for all the _print_ libdevice functions in the latest version. 

@eellison How about the actual codegen static methods? https://github.com/pytorch/pytorch/blob/main/torch/_inductor/codegen/triton.py#L769 To be more consistent, I think we will need the same upcast on inputs.

blaine-rister commented on test/inductor/test_torchinductor.py:
```diff
@@ -11808,6 +11809,20 @@ def f(x, mask):
+        def test_codegen_upcast_to_fp32(self, upcast_to_fp32):
+            @torch.compile
+            def func(a, b):
+                return a * b
```
Since this PR overrides a lot of other ops like `tan`, `cos`, etc., we might want to check that they also support fp16. Could we parametrize the test for other op types?
```
@parametrize("upcast_to_fp32", [False, True])
@parametrize("op", [torch.sum, torch.mul, torch.cos]) # Expand to all ops modified in this PR
 def test_codegen_upcast_to_fp32(self, op: Callable upcast_to_fp32: bool):
       @torch.compile
        def func(a, b):
            return a * b

blaine-rister reviewed (approved):
Nice work! Thanks for this important feature which will let us use more precise dtypes.

arui-meta committed:
[Inductor/Triton] Customize triton codegen to optionally upcast to float32 (#132406)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/132406

- add inductor config field `triton_codegen_upcast_to_fp32`

Test Plan:
```
AFG_DUMP_DBG_IR_STAGES=1 AFG_USE_CPUB=1 AFG_DUMP_NNMODEL_EXEC_MTIMER=1 buck2 run mode/{opt,inplace,mtia} //mtia/compiler/graph_compiler/test_library:afg_inductor_test -- -k tanh

```

Reviewed By: eellison, blaine-rister

Differential Revision: D60536337
